### ISSUE TITLE
TOC scroll position jumps to top after clicking a link on long TOCs (RTD theme)

### ISSUE THREAD

**Issue by sbutcher** [mkdocs/mkdocs#803]
Summary: Using ReadtheDocs theme, when your TOC is long and exceeds the browser height and a scrollbar is required, if you scroll down and click on a link, the scroll position jumps to the top again after opening your page. 

Intended behaviour: the position in the TOC is remembered for you to continue clicking through pages near the ones you are currently reading.

Example: go to http://sandboxdocs.readthedocs.org/ , and reduce your browsers height until you see the scrollbar. scroll down the left hand menu and choose one of the last links e.g. Web Service api.
You will note that instead of retaining the position in the TOC, the position has now jumped to the top again.

**Comment by waylan** [comment]
Is there even any elegant way to resolve this? When a link is clicked, the entire page is reloaded, including the TOC. 

True, there are solutions like using frames (shudder) or AJAX. However, in each of those, the page is served in pieces and only the individual pieces are loaded. As MkDocs is a "static page generator" its primary feature is that each page is a static, complete page on the server. No piecemeal assembling of pages when the request is made. Therefore, the obvious solution is in direct opposition to the main feature of MkDocs. 

I feel like in the past we wouldn't even consider such a request. However, today, with the prevalence of non-static pages (especially single-page apps) on many web sites, this behavior has become "normal" and that is the only reason this wasn't rejected immediately. So how can we "fake it?"

I suppose a URL query string could work (`http://example.com/some/page?toc=5`), but do we really want that? First, that makes for ugly URLs. Then, some JavaScript would be required to scroll the TOC to position 5 (whatever that might mean).

Sure, you could use the URL hash/fragment (`http://example.com/some/page#someid`), but that should be reserved for the position on the main body of the page (rather than the TOC).

I suppose JavaScript could write data to a cookie (or some client side DB) and retrieve it later. But that feels wrong for a "static site" to me. And then when would you update the info? Constantly as the TOC is scrolled or only when a TOC link is clicked?

I'm not aware of any other ways to maintain information across pages which don't require more sophisticated server setups than "static servers."

**Comment by d0ugal** [comment]
The information we need is already there - no need for cookies etc. The JavaScript on the ReadTheDocs theme would just need to scroll the left nav to the top of `li.toctree-l1.current`

**Comment by d0ugal** [comment]
This hacky line isn't too far off, but the offset is wrong. I'm not sure how we get the offset within the div.

```JavaScript
$('.wy-nav-side').scrollTop($('li.toctree-l1.current').offset().top);
```

**Comment by d0ugal** [comment]
This is closer and I think would be fine if it is only ever ran on page load.

```JavaScript
$('.wy-nav-side').scrollTop($('li.toctree-l1.current').offset().top - $('.wy-nav-side').offset().top);
```

**Comment by waylan** [comment]
Thanks @d0ugal for reminding me how old I am. Anyway, glad I was wrong out this.

**Comment by d0ugal** [comment]
You are welcome :)

I reclassified this as an Enhancement, it was never really a bug. That is how browsers should work, but this would be a nice touch to make it a bit better.

**Comment by marcelstoer** [comment]
I played with @d0ugal's latest snippet. I works quite well an reliably, thanks. However, the UX is odd for me, it doesn't feel natural at all.
1. I think it should only scroll if the selected item (or the submenu it opens) exceeds the viewport i.e. isn't fully visible.
2. It shouldn't scroll all the way to the top but just enough to balance it in the middle of the viewport height.

**Comment by d0ugal** [comment]
@marcelstoer Yeah, that's fair - it should probably only do it if the screen height is smaller than the offset calculated in the code I added.

**Comment by sbutcher** [comment]
Excellent, thanks for this. i offset this slightly to improve look and feel, so the line in my extra.js is
~~~~~~~~~~~~~~~
 $('.wy-nav-side').scrollTop($('li.toctree-l1.current').offset().top - $('.wy-nav-side').offset().top - 80);
~~~~~~~~~~~~~~~
to avoid hiding the search box, but i use a smaller search box section, so others may require something larger than 80.

**Comment by seanmadsen** [comment]
Below is the change I made to workaround this issue. It improves aforementioned workarounds by only performing the auto-scroll when necessitated by the active menu item being at least partially off-screen. I find this change overall to offer a very intuitive user experience. 

added the following to `mkdocs.yml`:

```yaml
extra_javascript:
  - js/custom.js
```
Created `js/custom.js` as follows: 

```javascript
$(function() {
  $.fn.isFullyWithinViewport = function(){
      var viewport = {};
      viewport.top = $(window).scrollTop();
      viewport.bottom = viewport.top + $(window).height();
      var bounds = {};
      bounds.top = this.offset().top;
      bounds.bottom = bounds.top + this.outerHeight();
      return ( ! (
        (bounds.top <= viewport.top) ||
        (bounds.bottom >= viewport.bottom)
      ) );
  };
  if( !$('li.toctree-l1.current').isFullyWithinViewport() ) {
    $('.wy-nav-side')
      .scrollTop(
        $('li.toctree-l1.current').offset().top -
        $('.wy-nav-side').offset().top -
        60
      );
  }
});
```

**Comment by cgwyllie** [comment]
In the readthedocs theme we're seeing this line `bounds.top = this.offset().top;` in the snippet posted by @seanmadsen throwing an exception on some page loads.

Seems this happens when the currently active TOC item is below the fold (scrolled off screen).

Here's the stack trace:

```
Uncaught TypeError: Cannot read property 'top' of undefined
    at n.fn.init.$.fn.isFullyWithinViewport (theme.js:67)
    at HTMLDocument.<anonymous> (theme.js:74)
    at j (jquery-2.1.1.min.js:2)
    at Object.fireWith [as resolveWith] (jquery-2.1.1.min.js:2)
    at Function.ready (jquery-2.1.1.min.js:2)
    at HTMLDocument.I (jquery-2.1.1.min.js:2)
```

As this was merged into the base theme in #1156 we've had to override the theme.js file to remove the scroll behaviour as we already had something similar in our theme's custom.js:

```
// Scroll active menu item into view
try {
    var offset = $('nav li.current:first')[0].offsetTop;
    // Somewhat arbitrary cutoff point
    if (offset > 300) {
        $('nav li.current:first')[0].scrollIntoView();
    }
}
catch (e) { /* noop */ }
```

**Comment by waylan** [comment]
@cgwyllie if you are reporting a bug in the current code, please open a separate issue for that. This issue relates to an enhancement, which is less of a priority than a bug. Your report will easily get lost and forgotten as a comment here. As a separate issue, we can track it and mark it as fixed when it is.

**Comment by cgwyllie** [comment]
@waylan no problem, issue filed: https://github.com/mkdocs/mkdocs/issues/1177
